### PR TITLE
improved error handling for invalid association values #2

### PR DESCRIPTION
**Pull request by flip111**
Possibly to do:
1. Make custom Exception for line 713
2. Make custom Exception for line 817
3. Does the object check on line 816 slow down the code too much? Alternatively a try-catch could be put around line 1415 or higher up.

Previous: https://github.com/doctrine/doctrine2/pull/1016

### CODE REVIEW

File: lib/Doctrine/ORM/UnitOfWork.php [hunk 1]
```diff
@@ -707,7 +707,11 @@ public function computeChangeSet(ClassMetadata $class, $entity)
                 continue;
             }
 
-            $this->computeAssociationChanges($assoc, $val);
+            try {
+                $this->computeAssociationChanges($assoc, $val);
+            } catch (\Exception $ex) {
```
Comment (Ocramius) [inline lib/Doctrine/ORM/UnitOfWork.php]: Don't catch a generic exception here: be specific

File: lib/Doctrine/ORM/UnitOfWork.php [hunk 2]
```diff
@@ -809,6 +813,10 @@ private function computeAssociationChanges($assoc, $value)
         $targetClass    = $this->em->getClassMetadata($assoc['targetEntity']);
 
         foreach ($unwrappedValue as $key => $entry) {
+            if (! ($entry instanceof $assoc['targetEntity']))) {
```
Comment (Ocramius) [inline lib/Doctrine/ORM/UnitOfWork.php]: instanceof `$targetClass->name`

File: lib/Doctrine/ORM/UnitOfWork.php [hunk 3]
```diff
@@ -809,6 +813,10 @@ private function computeAssociationChanges($assoc, $value)
...
+                throw ORMInvalidArgumentException::invalidAssociation($entry);
```
Comment (Ocramius) [inline lib/Doctrine/ORM/UnitOfWork.php]: You should throw with `$targetClass`, `$assoc`, the original entity metadata and the `$entry`: that allows you to build a more specific error message.

File: lib/Doctrine/ORM/ORMInvalidArgumentException.php [hunk 4]
```diff
@@ -198,4 +198,20 @@ private static function objToStr($obj)
     {
         return method_exists($obj, '__toString') ? (string)$obj : get_class($obj).'@'.spl_object_hash($obj);
     }
+
+    /**
+     * @return ORMInvalidArgumentException
+     */
+    public static function invalidAssociation($relation, $fieldname, $value) {
```
Comment (Ocramius) [inline lib/Doctrine/ORM/ORMInvalidArgumentException.php]: CS: brace should be on a new line

File: lib/Doctrine/ORM/ORMInvalidArgumentException.php [hunk 5]
```diff
@@ -198,4 +198,20 @@ private static function objToStr($obj)
...
+        return new self('Expected an Object for relation '.get_class($relation).'::'.$fieldname.' got '.gettype($value).' instead.');
+    }
+
+    /**
+     * @return ORMInvalidArgumentException
+     */
+    public static function invalidAssociation($entry) {
```
Comment (Ocramius) [inline lib/Doctrine/ORM/ORMInvalidArgumentException.php]: CS: brace should be on a new line

File: lib/Doctrine/ORM/ORMInvalidArgumentException.php [hunk 6]
```diff
@@ -198,4 +198,20 @@ private static function objToStr($obj)
...
+
+    /**
+     * @return ORMInvalidArgumentException
+     */
+    public static function invalidAssociation($relation, $fieldname, $value) {
+        return new self('Expected an Object for relation '.get_class($relation).'::'.$fieldname.' got '.gettype($value).' instead.');
+    }
+
+    /**
+     * @return ORMInvalidArgumentException
```
Comment (Ocramius) [inline lib/Doctrine/ORM/ORMInvalidArgumentException.php]: `@return self`

File: lib/Doctrine/ORM/ORMInvalidArgumentException.php [hunk 7]
```diff
@@ -198,4 +198,20 @@ private static function objToStr($obj)
     {
         return method_exists($obj, '__toString') ? (string)$obj : get_class($obj).'@'.spl_object_hash($obj);
     }
+
+    /**
+     * @return ORMInvalidArgumentException
```
Comment (Ocramius) [inline lib/Doctrine/ORM/ORMInvalidArgumentException.php]: `@return self`

Comment (Ocramius) [inline lib/Doctrine/ORM/ORMInvalidArgumentException.php]: Missing parameters in the docblock

File: lib/Doctrine/ORM/ORMInvalidArgumentException.php [hunk 8]
```diff
@@ -198,4 +198,20 @@ private static function objToStr($obj)
...
+     */
+    public static function invalidAssociation($relation, $fieldname, $value) {
+        return new self('Expected an Object for relation '.get_class($relation).'::'.$fieldname.' got '.gettype($value).' instead.');
```
Comment (Ocramius) [inline lib/Doctrine/ORM/ORMInvalidArgumentException.php]: Spaces around dots: also use `sprintf()` instead

File: lib/Doctrine/ORM/ORMInvalidArgumentException.php (hunk 6 above)
Comment (Ocramius) [inline lib/Doctrine/ORM/ORMInvalidArgumentException.php]: Missing docblock parameters

File: lib/Doctrine/ORM/ORMInvalidArgumentException.php [hunk 9]
```diff
@@ -198,4 +198,20 @@ private static function objToStr($obj)
...
+    }
+
+    /**
+     * @return ORMInvalidArgumentException
+     */
+    public static function invalidAssociation($entry) {
+        $ex = new self(gettype($entry) . ' is not an Object.');
+        $ex->value = $entry;
```
Comment (Ocramius) [inline lib/Doctrine/ORM/ORMInvalidArgumentException.php]: Don't set a dynamic property: simply limit yourself to inlining it *IF* it is a scalar type

**Comment by Ocramius** [comment]
@flip111 also, please provide tests for this enhancement

File: lib/Doctrine/ORM/ORMInvalidArgumentException.php (hunk 9 above)
Comment (stof) [inline lib/Doctrine/ORM/ORMInvalidArgumentException.php]: or use ``self::objToStr`` for objects

File: lib/Doctrine/ORM/ORMInvalidArgumentException.php (hunk 4 above)
Comment (stof) [inline lib/Doctrine/ORM/ORMInvalidArgumentException.php]: and public methods should be before the private one

File: lib/Doctrine/ORM/UnitOfWork.php (hunk 1 above)
Comment (flip111) [inline lib/Doctrine/ORM/UnitOfWork.php]: Not sure how i can track all the possible situations which yield all possible Exceptions

File: lib/Doctrine/ORM/ORMInvalidArgumentException.php (hunk 9 above)
Comment (flip111) [inline lib/Doctrine/ORM/ORMInvalidArgumentException.php]: The exception is only thrown when the variable is not an object

File: lib/Doctrine/ORM/ORMInvalidArgumentException.php (hunk 8 above)
Comment (flip111) [inline lib/Doctrine/ORM/ORMInvalidArgumentException.php]: trend break with other methods of same class

File: lib/Doctrine/ORM/ORMInvalidArgumentException.php (hunk 7 above)
Comment (flip111) [inline lib/Doctrine/ORM/ORMInvalidArgumentException.php]: `@return self` trend break with other methods of same class

File: lib/Doctrine/ORM/ORMInvalidArgumentException.php (hunk 6 above)
Comment (flip111) [inline lib/Doctrine/ORM/ORMInvalidArgumentException.php]: `@return self` trend break with other methods of same class

**Comment by flip111** [comment]
@Ocramius i totally lost track (since 10 months) how i even arrived at this situation where this error would be useful. I hope not to forgot the PR this time though ^^

**Comment by Ocramius** [comment]
@flip111 the original PR was probably trying to give a better meaning to stuff like:

```php
class Foo
{
    /** @ManyToOne(targetEntity="Bar") */
    public $bar;
}

$foo = new Foo;

$foo->bar = 'baz'; // wrong - exception
$foo->bar = []; // wrong - exception
$foo->bar = new ArrayCollection; // wrong - exception
$foo->bar = new stdClass; // wrong - exception
$foo->bar = new Bar; // correct
```

File: lib/Doctrine/ORM/UnitOfWork.php (hunk 1 above)
Comment (Ocramius) [inline lib/Doctrine/ORM/UnitOfWork.php]: @flip111 writing these tests isn't that problematic if you keep the test asset as simple as possible.

**Comment by fprochazka** [comment]
Oh boy, I would strongly discourage you from writing code in github editor. Checkout the branch and fix it locally. It's more effective and IDE tells about problems github won't.

**Comment by Ocramius** [comment]
This was manually rebased/tested and merged into `master` at doctrine/doctrine@39766e645cd5c332cff2d6a24b9eba5656c05697

Thanks @flip111!

**Comment by flip111** [comment]
Thank you @Ocramius i had already a hard time imagining being able to finish this PR

**Comment by Ocramius** [comment]
@flip111 check the merge commit, the test was not that terrible after all :-)

**Comment by flip111** [comment]
Ye i know it's not a big change. Just that Doctrine internal code is pretty hard to get into. I guess that goes for any ORMs though..

I checked the commit, you added two cases on the 2000+ lines, that's nice. I remember when i was testing i needed that catch around line 710 ... perhaps that has changed by now.

**Comment by Ocramius** [comment]
@flip111 yeah, the association exception throwing was enough :-)

**Pull request closed without merging**